Automattic/wp-calypso
# Update jsdoc eslint plugin

#### Proposed Changes
Separating this from #72245 to simplify things. Updates eslint-plugin-jsdoc across the repo to properly support Node 18. cc @anomiex 

The initial update introduced two problems:
1. Firstly, JSDoc wanted to replace `Object` with `object`. 
2. Secondly, there were a lot of issues with require-returns check. It seemed to have gotten a lot smarter :)

The first fix is easy -- just add a setting which allows both Object and object. I don't see a big reason to enforce one over the other at this point, especially since we'd rather use Typescript.

The second one just meant going through several files and fixing the return type. Almost every error was caused by functions like this:

```js
/**
 * @returns {string}
 */
function foo( x ) {
  if ( x ) {
    return "bar";
  }
}
```

This is obviously problematic, because the real return type is `string|undefined`. So I just made that adjustment in several files.

There are a few more issues reported by the main linter build below, but not enough to fail the build. I did find weird behavior in the eslint jsdoc lib, which I reported here: https://github.com/gajus/eslint-plugin-jsdoc/issues/949

#### Testing Instructions
CI, and verify that changes make sense. The full eslint build is running here: https://teamcity.a8c.com/buildConfiguration/calypso_CheckCodeStyle/9329626
